includes/Core/Util/Site_Health_Status.php

```diff
@@ -0,0 +1,284 @@
+				);
+
+				return $tests;
+			}
+		);
+	}
+
+	/**
+	 * Gets all REST routes.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return REST_Route[]
+	 */
+	private function get_rest_routes() {
+		return array(
+			new REST_Route(
+				'core/site/data/tags-placement-test',
+				array(
+					array(
+						'methods'             => WP_REST_Server::READABLE,
+						'callback'            => function() {
+							return $this->tags_placement_test();
+						},
+						'permission_callback' => function () {
+							return current_user_can( Permissions::SETUP );
+						},
+					),
+				)
+			),
+		);
+	}
+
+	/**
+	 * Checks if the modules tags are placed on the website.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string
+	 */
+	protected function tags_placement_test() {
+		global $wp_version;
+
+		$result = array(
+			'label'   => __( 'Tags Placement', 'google-site-kit' ),
+			'status'  => 'recommended',
+			'badge'   => array(
+				'label' => __( 'Site Kit', 'google-site-kit' ),
+				'color' => 'blue',
+			),
+			'actions' => '',
+			'test'    => 'tag_placement',
+		);
+
+		if ( version_compare( $wp_version, '5.6', '<' ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'This feature requires WordPress version 5.6 or higher', 'google-site-kit' )
+			);
+
+			return $result;
+		}
+
+		$response = wp_remote_get( site_url() ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_remote_get_wp_remote_get
+
+		if ( is_wp_error( $response ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'There was an error while trying to get the status, please try again later.', 'google-site-kit' )
+			);
+
+			return $result;
+		}
+
+		$response = wp_remote_retrieve_body( $response );
+
+		$active_modules = $this->get_active_modules();
+
+		if ( empty( $active_modules ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'Tag status not available: AdSense, Tag Manager, and Analytics modules are not connected.', 'google-site-kit' )
+			);
+
+			return $result;
+		}
+
+		$description = array();
+
+		foreach ( $active_modules as $module ) {
+			$tag_found = $this->check_if_tag_exists( $module->slug, $response );
+
+			if ( $tag_found ) {
+				$description[] = $tag_found;
+			}
+		}
+
+		if ( ! empty( $description ) ) {
+			$result['description'] = '<ul>' . join( "\n", $description ) . '</ul>';
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Gets active modules filtered to account only for
+	 * Analytics, AdSense and Tag Manager.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Filtered active modules instances.
+	 */
+	protected function get_active_modules() {
+		$active_modules = $this->modules->get_active_modules();
+
+		$active_modules = array_filter(
+			$active_modules,
+			function( $module ) {
+				return in_array(
+					$module->slug,
+					array(
+						Analytics_4::MODULE_SLUG,
+						AdSense::MODULE_SLUG,
+						Tag_Manager::MODULE_SLUG,
+					),
+					true
+				);
+			}
+		);
```

Comment:
The core class should know nothing about modules, it should only operate with entities available in the Core namespace. I think we need to rework your approach a bit:

1. Add a new interface `Module_With_Tag` to the `Core/Modules` namespace that has two methods: `register_tag()` and `has_tag( $content )`.
1. Add the new `Module_With_Tag` interface to `AdSense`, `Analytics_4` and `Tag_Manager` module classes.
1. Implement a new `has_tag( $content )` method in those classes that creates an appropriate tag matcher instance and return the result of the `has_tag` method call.
1. In this `get_active_modules ` method return only those modules that implement the `Module_With_Tag` interface.
1. Update the `check_if_tag_exists` method to accept a module instance instead of a module slug and call `$module->has_tag( ... )` method to determine whether the tag is added for that module.

Reply:
I expanded on this approach, and made `Module_With_Tag_Trait` to avoid repeating the code and make it more centralised